### ISSUE TITLE
Add editor font support to RStudio Server

### ISSUE THREAD

**Issue by statkclee** [rstudio/rstudio#2534]
### System details

    RStudio Edition : 
    RStudio Version : 1.1.442
    OS Version      :  Ubuntu 16.04
    R Version       :  3.4.4 (2018-03-15)

### Steps to reproduce the problem
Select menu from Tools -> Global Options -> Appearance

### Describe the problem in detail

There is no "Editor Font" dropdown UI.
The default font is not adequate as the main CJK editor font.

### Describe the behavior you expected

It should show "Editor Font" dropdown UI just like RStudio Desktop version.

![image](https://user-images.githubusercontent.com/5325425/38072571-e90e2992-3361-11e8-894d-d5cfe3508044.png)

**Comment by jmcphers** [comment]
The reason there isn't an Editor Font dropdown in RStudio Server is that web applications don't get to ask your system for a list of installed fonts. A further complication is that you can connect to RStudio Server from several different computers with different sets of fonts, so the font you chose on one system might not be available on another one.

We might be able to offer a font selection experience of some kind, such as (a) testing against a large list of known monospace fonts, (b) providing a free-form font name entry box, and/or (c) providing a place to upload a monospace web font of your choice, but it is not possible to fully replicate the desktop experience.

Appreciate the suggestion!

**Comment by statkclee** [comment]
How about embedding one Korean coding fonts?
From my experience so far, there are a few Korean coding font. One of them is Naver D2 Coding font.

- https://github.com/naver/d2codingfont

Such that instead of supporting all the customized fonts for R Coding, only one coding font for each country will be sufficient enough.

**Comment by balwierz** [comment]
What about having a free-form _preferred_ font name(s)?
Rstudio server could try using them and if not available use the default font.

I am using fira-code fonts for nice ligatures for <- != >= && and so on. When I force firefox to use it as a sheriff-less font, it does not apply ligatures. (Rstudio desktop does.)

**Comment by jmcphers** [comment]
See also https://github.com/rstudio/rstudio/issues/4134 (Use embedded webfonts for primary UI and fixed width fonts)

**Comment by loudermilk** [comment]
Would make my RStudio experience better to be able to use Fira Code ligatures - https://github.com/tonsky/FiraCode

**Comment by jiho** [comment]
What about choosing from web font services (e.g. Google Fonts)?

**Comment by Gabryxx7** [comment]
Hi everyone! I have RStudio-Server installed on my own server and I decided to style it up a little bit with custom themes and fonts. I came across this issue and I decided to give my own solution which works rather well and it's extremely easy (I am gonna use Fire Code for the example @loudermilk, but it would work with google fonts as well @jiho or whatever server) :

1. (actually optional but some people might prefer it) Download Fira Code and put it somewhere accessible from the internet (your github.io page, your own server, or even just a link to dropbox, or  even google drive or google font, as long as it's a direct link to the file!). I have it on my own server, in the same folder from where my websites gets css/js files, so accessing `http://test.com/fonts/Fira/ttf/FiraCode-Light.ttf` would download the .ttf file. But for this example I will get the files raw from the Fira Code GitHub repository:`https://github.com/tonsky/FiraCode/raw/master/distr/ttf/FiraCode-Light.ttf`
1. Find a way to apply custom CSS, I am using the "User JavaScript and CSS" extension for Chrome
2. Download Fira Code from GitHub, and open the .css file that comes with it with any editor (I use Visual Studio Code)
3. Modify the urls so that point to your Fira folder on the server: I ran a simple find and replace, I looked for `url('` and replaced with `url('https://github.com/tonsky/FiraCode/raw/master/distr/`.
4. Edit your Chrome custom CSS (the extension I use allows me to specify different CSS or even JS code for different domains) and add one last rule so that the final CSS loaded everytime a page on the domain is opened would look something like"

```css
@font-face{
    font-family: 'Fira Code';
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Light.eot');
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Light.eot') format('embedded-opentype'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff2/FiraCode-Light.woff2') format('woff2'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff/FiraCode-Light.woff') format('woff'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/ttf/FiraCode-Light.ttf') format('truetype');
    font-weight: 300;
    font-style: normal;
}

@font-face{
    font-family: 'Fira Code';
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Regular.eot');
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Regular.eot') format('embedded-opentype'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff2/FiraCode-Regular.woff2') format('woff2'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff/FiraCode-Regular.woff') format('woff'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/ttf/FiraCode-Regular.ttf') format('truetype');
    font-weight: 400;
    font-style: normal;
}

@font-face{
    font-family: 'Fira Code';
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Medium.eot');
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Medium.eot') format('embedded-opentype'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff2/FiraCode-Medium.woff2') format('woff2'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff/FiraCode-Medium.woff') format('woff'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/ttf/FiraCode-Medium.ttf') format('truetype');
    font-weight: 500;
    font-style: normal;
}

@font-face{
    font-family: 'Fira Code';
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Bold.eot');
    src: url('https://github.com/tonsky/FiraCode/raw/master/distr/eot/FiraCode-Bold.eot') format('embedded-opentype'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff2/FiraCode-Bold.woff2') format('woff2'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/woff/FiraCode-Bold.woff') format('woff'),
         url('https://github.com/tonsky/FiraCode/raw/master/distr/ttf/FiraCode-Bold.ttf') format('truetype');
    font-weight: 700;
    font-style: normal;
}

.ace_layer.ace_text-layer {
    font-family: 'Fira Code' !important;
}
```
Et voilà! Chrome, the extension and CSS will take care of downloading the CSS and all the required fonts files and apply the new CSS to the pages of the domain!

Proof of working Fira Code on RStudio-Server

![Screen Shot 2019-04-25 at 6 25 08 pm](https://user-images.githubusercontent.com/6282055/56721084-7c969500-6787-11e9-8051-8868f7ed12fd.png)

One could also edit the Rstudio-Server main PHP/HTML file (not sure what language is used), and add a line on top to include the custom css with your fonts, this would be a pretty good solution, but it wouldn't allow different users to have different fonts.

*TL;DR Download che Chrome extension "User JavaScript and CSS" and just open your rstudio-server on chrome, copy-paste the above CSS code and save*

**Comment by taiyodayo** [comment]
Thank you Gabriele san, it is plain fantastic!!
Never thought it could be achieved so elegantly.
R coders in my team is forever thankful for your suggestion!

**Comment by antoine-sachet** [comment]
The CSS given by @Gabryxx7 is now broken (`/eot` folder has been removed) but in any case, Fira Code is now available via the `jsdelivr` CDN (see [README](https://github.com/tonsky/FiraCode#browser-support).

So the CSS can simply be:
```css
@import url(https://cdn.jsdelivr.net/gh/tonsky/FiraCode@1.206/distr/fira_code.css);

.ace_layer.ace_text-layer {
    font-family: 'Fira Code' !important;
}
```

**Comment by Gabryxx7** [comment]
You could actually store the FiraCode files in your own server and use the URL in your CSS, you would have to update it yourself, but if you have `git` in your server you can just refer to the repository folder on your server.

**Comment by y1zhou** [comment]
> However there is an issue with the cursor position. It does not fall between letters and does not reach the end of the line, so it is pretty unusable.

@antoine-sachet This was the case when I changed the font to Fira Code Retina. My solution was declaring the font size in the custom CSS:

```css
.ace_layer.ace_text-layer {
        font-size: 13px !important;
        font-family: 'Fira Code Retina', monospace !important;
    }
}
```

You may have to play with the values. `Tools - Global options - Appeariance - Editor font size` is 10 in my settings.

**Comment by eliocamp** [comment]
Can confirm that @antoine-sachet fix was unusable for me. :(

**Comment by antoine-sachet** [comment]
@eliocamp The solution from @y1zhou worked for me. My font size is 11 and with my screen resolution `12.7px` is perfect.

```CSS
@import url('https://fonts.googleapis.com/css?family=Fira+Code');

.ace_layer.ace_text-layer {
    font-family: 'Fira Code Retina', monospace !important;
    font-size: 12.7px !important;
}
```

**Comment by eliocamp** [comment]
Yay! That precise combination seems to work fairly well (there's still some small shift in some cases, but workable) 
Thanks!

**Comment by leonardonormando** [comment]
Thank you all! I'm using font size 11 (screen resolution 1920 x 1080) and the following code works for me:

`
@import url('https://fonts.googleapis.com/css?family=Fira+Code&display=swap');

.ace_layer.ace_text-layer {
    font-family: 'Fira Code' !important;
    font-size: 13.9px !important;
}
`

**Comment by ferroao** [comment]
Should the workaround affect the font of the terminal? I see no change in it.

**Comment by Gabryxx7** [comment]
> Should the workaround affect the font of the terminal? I see no change in it.

The console's CSS id is `#rstudio_console_output` while the terminal's classes are `terminal xterm xterm-theme-default xterm-cursor-style-block xterm-cursor-blink ace_editor GD15MCFCAW` So you should add the id or the classes you need to your CSS selector.

**Comment by hyiltiz** [comment]
How is it closed (but re-opened)? I still do not see the font support, and would rather not have to dig into server css code, or use a browser CSS injection addon.

**Comment by gtritchie** [comment]
This feature was added to RStudio 1.4, still under development. It is available in the daily builds: https://dailies.rstudio.com/

**Comment by hyiltiz** [comment]
How long does it take in general for a daily test builds to reach release?

**Comment by Thell** [comment]
A fairly recent [community post](https://community.rstudio.com/t/command-palette-from-daily-coming-to-preview-soonish/73808/2) in response to a feature showing up in previews...

> No, not in the near future; the Command Palette is part of the 1.4 release and won't be backported to the 1.3 branch. **1.4 previews will probably start appearing around September or October.**
(emphasis added)

🤞

**Comment by hyiltiz** [comment]
Now if we the time series about when a dev version shows up in preview and will be ready for release, we could build a model to predict when the said 1.4 would hit Release (and thus actually close this Issue for good).

**Comment by ronblum** [comment]
Hi all—just to clarify the distinction between daily builds and preview builds: 
- The [RStudio daily development builds](https://dailies.rstudio.com/) include all of  the code as it is checked in, on a daily basis. This is the  current stage for 1.4 Wax Begonia.
- We post preview builds later in the development of cycle, after the new features and updates are complete, and we've 
 conducted some testing. We continue with the daily releases, as well. 

So when a new feature is checked in, it should be available in the daily builds in the next day or two. Please note that both the daily and preview builds are intended for testing purposes only.

**Comment by msberends** [comment]
Just for future searchers: you can now add fonts to `/etc/rstudio/fonts` and then change the ‘Editor font’ using the UI.


So I’m just pasting the contents of `/etc/rstudio/fonts/README.md` here, I’m on RStudio Server 2022.07.0 on Ubuntu Server 22.04 LTS:

```
This directory structure would make two fonts available, *Coding-Font* and
*Coding Font Two*.

Some fonts come in many different weights and styles. If you want these weights
and styles to be treated as single font, you can place them underneath a single
folder. This is useful when a theme uses bold or italic variants of a font to
decorate code (e.g., to set comments in italics).

To do this, create subfolders with the font's weight or style. For example,
this creates a single font, "Coding Font 3", which has two weights (400 and 700
for regular and bold, respectively) and an italic style for each weight.

    + fonts/
    |
    +-- Coding Font Three/
        |
        +-- 400/
        |   |
        |   +-- CodingFont3-Regular.woff
        |   |
        |   +-- italic/
        |       |
        |       +-- CodingFont3-Italic.woff
        |
        +-- 700/
            |
            +-- CodingFont3-Bold.woff
            |
            +-- italic/
                |
                +-- CodingFont3-BoldItalic.woff
```

Works great for Fira Code:

![84DD31DF-834F-4AD2-B669-CE2A27A6FC7A](https://user-images.githubusercontent.com/31037261/180660183-00f6e4f6-f500-4a67-b7f8-74d059a18c7b.jpeg)

**Comment by jmcphers** [comment]
More documentation on this feature in the admin guide here: https://docs.rstudio.com/ide/server-pro/r_sessions/customizing_session_settings.html#fonts
